geniusyield-market-maker/src/GeniusYield/MarketMaker/Utils.hs

```diff
@@ -53,3 +66,132 @@ instance HasPartialOrderNftScript DEXInfo where
 
 instance HasPartialOrderConfigAddr DEXInfo where
   getPartialOrderConfigAddr = dexPartialOrderConfigAddr
+
+
+-------------------------------------------------------------------------------
+-- Common Price Resolution
+-------------------------------------------------------------------------------
+
+-- | Generic time resolution for OHLC Candles
+data CommonResolution = CRes5m | CRes15m | CRes30m | CRes1h | CRes4h
+                      | CRes1d | CRes1w  | CRes1mo
+                      deriving stock (Eq, Ord, Show, Generic)
+
+instance FromJSON CommonResolution
+
+class PriceResolution a where
+  resolutionInherited :: Map.Map CommonResolution a
+
+instance PriceResolution Maestro.Resolution where
+  resolutionInherited = Map.fromList [ (CRes5m, Res5m), (CRes15m, Res15m), (CRes30m, Res30m), (CRes1h, Res1h)
+                                     , (CRes4h, Res4h), (CRes1d, Res1d), (CRes1w, Res1w), (CRes1mo, Res1mo) ]
+
+fromCommonResolution :: PriceResolution a => CommonResolution -> IO a
+fromCommonResolution = maybe (throwIO $ userError "Undefined common price resolution.") pure
+                       . flip Map.lookup resolutionInherited
+
+
+-------------------------------------------------------------------------------
+-- TapTools OHLCV query
+-------------------------------------------------------------------------------
+
+-- | Taptools time resolution for OHLC Candles
+data TtResolution = TtRes3m | TtRes5m | TtRes15m | TtRes30m | TtRes1h | TtRes2h | TtRes4h | TtRes12h
+                  | TtRes1d | TtRes3d | TtRes1w  | TtRes1mo
+                  deriving stock (Eq, Ord, Generic)
+
+instance FromJSON TtResolution
+
+instance Show TtResolution where
+  show = let kvm = Map.fromList [ (TtRes3m, "3m"), (TtRes5m, "5m"), (TtRes15m, "15m"), (TtRes30m, "30m")
+                                , (TtRes1h, "1h"), (TtRes2h, "2h"), (TtRes4h, "4h"), (TtRes12h, "12h")
+                                , (TtRes1d, "1d"), (TtRes3d, "3d"), (TtRes1w, "1w"), (TtRes1mo, "1M") ]
+         in  fromJust . flip Map.lookup kvm
+
+instance PriceResolution TtResolution where
+  resolutionInherited = Map.fromList [ (CRes5m, TtRes5m), (CRes15m, TtRes15m), (CRes30m, TtRes30m), (CRes1h, TtRes1h)
+                                     , (CRes4h, TtRes4h), (CRes1d, TtRes1d), (CRes1w, TtRes1w), (CRes1mo, TtRes1mo) ]
+
+type TtUnit = String; type TtInterval = String
+
+type TtAPI =
+  "token" :> "ohlcv" :> QueryParam "unit" TtUnit
+                     :> QueryParam "interval" TtInterval
```

Comment:
isn't "interval" the `TtResolution`? Why use `String` here?

Reply:
What I get from your comment is that `QueryParam` is able to use the `ToJSON` instance to parse `TtResolution` as a string, and that therefore I should define the corresponding instance.  Then I could use directly `TtResolution` as an argument of `QueryParam`, instead of `TtInterval`.  Am I understanding you correctly?

Reply:
> What I get from your comment is that QueryParam is able to use the ToJSON instance...

No, please have a look at [this section on `FromHttpApiData`/`ToHttpApiData`](https://docs.servant.dev/en/stable/tutorial/Server.html#the-fromhttpapidata-tohttpapidata-classes).

Reply:
Done:  declared `instance ToHttpApiData TtResolution where` using pattern match and removed `TtInterval`.